keps/sig-network/0006-20180314-coredns-GA-proposal.md

```diff
@@ -0,0 +1,132 @@
+    * [User Cases](#use-cases)
+* [Unresolved Questions](#unresolved-questions)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+CoreDNS is sister CNCF project and is the successor to SkyDNS, on which kube-dns is based. It is a flexible, extensible
+authoritative DNS server and directly integrates with the Kubernetes API. It can serve as cluster DNS,
+complying with the [dns spec](https://git.k8s.io/dns/docs/specification.md). As an independent project,
+it is more actively developed than kube-dns and offers performance and functionality beyond what kube-dns has. For more details, see the [introductory presentation](https://docs.google.com/presentation/d/1v6Coq1JRlqZ8rQ6bv0Tg0usSictmnN9U80g8WKxiOjQ/edit#slide=id.g249092e088_0_181), or [coredns.io](https://coredns.io), or the [CNCF webinar](https://youtu.be/dz9S7R8r5gw).
+
+Currently, we are following the road-map defined [here](https://github.com/kubernetes/features/issues/427). CoreDNS is Beta in Kubernetes v1.10, which can be installed as an alternate to kube-dns.
+The purpose of this proposal is to graduate CoreDNS to GA and be the default DNS service, replacing the current implementation of kube-dns.
+
+## Motivation
+
+* CoreDNS is more flexible and extensible than what kube-dns. 
+* CoreDNS is easily extensible and maintainable using a plugin architecture.
+* CoreDNS has fewer moving parts than kube-dns, taking advantage of the plugin architecture, making it a single executable and single process.
+* It is written in Go, making it memory-safe (kube-dns includes dnsmasq which is not). 
+* CoreDNS has [better performance](https://github.com/kubernetes/community/pull/1100#issuecomment-337747482) than [kube-dns](https://github.com/kubernetes/community/pull/1100#issuecomment-338329100) in terms of greater QPS, lower latency, and lower memory consumption. 
+* It supports a number of [use cases](https://coredns.io/2017/05/08/custom-dns-entries-for-kubernetes/) that kube-dns doesn't. As a general-purpose authoritative DNS server it has a lot of functionality that kube-dns couldn't reasonably be expected to add.
+
+### Goals
+
+* Have CoreDNS run as the default DNS server both after a new installation and an upgrade of the existing installation.
+* Make CoreDNS available as an [image in gcr.io](https://github.com/kubernetes/kubernetes/pull/59753) and define a workflow/process to update the CoreDNS versions in the future.
+* Provide a kube-dns to CoreDNS upgrade path with configuration translation.
+* Retain the configuration of a CoreDNS deployment from Alpha (1.9) or Beta (1.10).
+
+### Non-Goals
+
+* Translation of CoreDNS ConfigMap back to kube-dns (i.e., downgrade).
+* Retaining the ConfigMap of a custom/modified image of CoreDNS during upgrade. That is, if the user has modified the Corefile in the ConfigMap, those changes will not be retained (the user should back it up and redo the changes on the new ConfigMap).
```

Comment:
Feels quite intrusive to overwrite user configuration on an upgrade without a path to prevent this. At least needs to be documented or warned for imo.

Reply:
Ok, I can document that and also make this statement a little clearer. 

Just to clarify, at upgrade time, the ConfigMap is not retained only if the coredns image we find is "unofficial" (not shipped with alpha (1.9) or beta (1.10)... ) and coredns will be upgraded to the latest image and default ConfigMap.

Reply:
Agree - this is user-hostile.  If you expose a ConfigMap you are publishing it as a mechanism to configure the app.  If you intend to blow it away, you should just not expose it. at all.

Now we have to talk about defaults vs user-overrides.

Reply:
Ok, but this is just for the ConfigMap, right? That is, how should we handle changes to the manifest/deployment itself?

In this case then we should add the new, updated default ConfigMap under a different name, to make it easy for the user to switch to that if they want to, or at least compare any differences. This also means that in the event of non-backwards compatible changes to the Corefile (which we might do, but only when we bump the minor - not patch - version), we will need to do something special.

@rajansandeep can you make sure the 1.0.x -> 1.1.x remains backwards compatible for the Corefile? If so we don't need to worry about that for a while.

Reply:
@johnbelamaric : what about this proposition for  upgrade 1-ConfigMap, 2-Deployment
1- never update the ConfigMap (as you suggest). It is created first time of installation of CoreDNS (either by upgrade from kube-dns, either by installation from scratch), and on subsequent upgrade, we keep that ConfigMap - Until we reach the issue of non-backward compatible corefile. At that time we figure out with special upgrade.

NOTE: does it worth to create a "coredns-default" ConfigMap ? It would show no more than the one of the coredns/deployment repository since kube-dns config no longer exist.

2- For the deployment part, first installation creates the deployment of CoreDNS. Subsequent upgrades only change the image of CoreDNS - unless admin prevent it by removing the label "kubeadm-auto-upgrade" that the first installation is setting.